core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java

```diff
@@ -334,6 +287,72 @@ public boolean isValid(String requestId) {
 		}
 	}
 
+	// Check SubjectConfirmation, at least one SubjectConfirmation must be valid
+	private void validateSubjectConfirmation(String responseInResponseTo) throws Exception {
+		final List<String> subjectConfirmationDataIssues = new ArrayList<>();
+		boolean validSubjectConfirmation = false;
+		NodeList subjectConfirmationNodes = this.queryAssertion("/saml:Subject/saml:SubjectConfirmation");
+		for (int i = 0; i < subjectConfirmationNodes.getLength(); i++) {
+			Node scn = subjectConfirmationNodes.item(i);
+
+			Node method = scn.getAttributes().getNamedItem("Method");
+			if (method != null && !method.getNodeValue().equals(Constants.CM_BEARER)) {
+				continue;
+			}
+
+			NodeList subjectConfirmationDataNodes = scn.getChildNodes();
+			for (int c = 0; c < subjectConfirmationDataNodes.getLength(); c++) {
+				if (subjectConfirmationDataNodes.item(c).getLocalName() != null && subjectConfirmationDataNodes.item(c).getLocalName().equals("SubjectConfirmationData")) {
+
+					Node recipient = subjectConfirmationDataNodes.item(c).getAttributes().getNamedItem("Recipient");
+					if (recipient == null) {
+						subjectConfirmationDataIssues.add("SubjectConfirmationData doesn't contain a Recipient");
+						continue;
+					}
+
+					if (!recipient.getNodeValue().equals(currentUrl)) {
+						subjectConfirmationDataIssues.add("SubjectConfirmationData doesn't match a valid Recipient");
+						continue;
+					}
+
+					Node inResponseTo = subjectConfirmationDataNodes.item(c).getAttributes().getNamedItem("InResponseTo");
+					if (inResponseTo != null && !inResponseTo.getNodeValue().equals(responseInResponseTo)) {
+						subjectConfirmationDataIssues.add("SubjectConfirmationData has an invalid InResponseTo value");
+						continue;
+					}
+
+
+					Node notOnOrAfter = subjectConfirmationDataNodes.item(c).getAttributes().getNamedItem("NotOnOrAfter");
+					if (notOnOrAfter == null) {
+						subjectConfirmationDataIssues.add("SubjectConfirmationData doesn't contain a NotOnOrAfter attribute");
+						continue;
+					}
+
+					DateTime noa = Util.parseDateTime(notOnOrAfter.getNodeValue());
+					if (noa.isEqualNow() || noa.isBeforeNow()) {
+						subjectConfirmationDataIssues.add("SubjectConfirmationData is no longer valid");
+						continue;
+					}
+
+					Node notBefore = subjectConfirmationDataNodes.item(c).getAttributes().getNamedItem("NotBefore");
+					if (notBefore != null) {
+						DateTime nb = Util.parseDateTime(notBefore.getNodeValue());
+						if (nb.isAfterNow()) {
+							subjectConfirmationDataIssues.add("SubjectConfirmationData is not yet valid");
+							continue;
+						}
+					}
+					validSubjectConfirmation = true;
+				}
+			}
+		}
+		if (!validSubjectConfirmation) {
+			String subjectConfirmationDataIssuesMsg = subjectConfirmationDataIssues.isEmpty() ? "" :
```

Comment:
When I implemented it I also though about add info about the reason why a SubjectConfirmation is rejected, but 
since is possible to have more than 1 SubjectConfirmationData element and I didn't found a way to differentiate the subjectConfirmationDataIssues (rather than mention the position returned on the NodeList ) I never implemented it.